openapi_python_client/parser/properties/union.py

```diff
@@ -67,16 +75,7 @@ def build(
                 return PropertyError(detail=f"Invalid property in union {name}", data=sub_prop_data), schemas
             sub_properties.append(sub_prop)
 
-        def flatten_union_properties(sub_properties: list[PropertyProtocol]) -> list[PropertyProtocol]:
-            flattened = []
-            for sub_prop in sub_properties:
-                if isinstance(sub_prop, UnionProperty):
-                    flattened.extend(flatten_union_properties(sub_prop.inner_properties))
-                else:
-                    flattened.append(sub_prop)
-            return flattened
-
-        sub_properties = flatten_union_properties(sub_properties)
+        sub_properties, discriminators_list = _flatten_union_properties(sub_properties)
```

Comment:
This flattening logic is a pre-existing behavior of the generator, causing it to treat this:
```yaml
  oneOf:
    - $ref: "#/components/schemas/TypeA"
    oneOf:
      - $ref: "#/components/schemas/TypeB"
      - $ref: "#/components/schemas/TypeC"
```

—as exactly equivalent to this:
```yaml
  oneOf:
    - $ref: "#/components/schemas/TypeA"
    - $ref: "#/components/schemas/TypeB"
    - $ref: "#/components/schemas/TypeC"
```

But now we have to also keep track of any discriminator definitions that might have existed in a nested union. For instance, it's totally valid to have something like this—
```yaml
  UnionType:
    oneOf:
      - oneOf:
        - $ref: "#/components/schemas/TypeA"
        - $ref: "#/components/schemas/TypeB"
        discriminator:
          property: prop1
      - oneOf:
        - $ref: "#/components/schemas/TypeC"
        - $ref: "#/components/schemas/TypeD"
        discriminator:
          property: prop2
```
In this example, TypeA and TypeB both have the property prop1, whose value will be "TypeA" or "TypeB"; but with TypeC and TypeD, the relevant property is prop2. The `discriminators_list` returned by `_flatten_union_properties` in this case looks like: `[DiscriminatorDefinition(property_name="prop1", value_to_model_map={"TypeA": <ModelProperty for TypeA>, "TypeB": <ModelProperty for TypeB>}),DiscriminatorDefinition(property_name="prop1", value_to_model_map={"TypeA": <ModelProperty for TypeA>, "TypeB": <ModelProperty for TypeB>}), DiscriminatorDefinition(property_name="prop2", value_to_model_map={"TypeC": <ModelProperty for TypeA>, "TypeD": <ModelProperty for TypeB>})]. The Jinja template logic will use this to build an appropriate piece of Python deserializer code where first it checks for prop1, then for prop2, etc.